src/XMakeTasks/AssemblyDependency/AssemblyFoldersFromConfig/AssemblyFoldersFromConfig.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Build.Shared;
+using ProcessorArchitecture = System.Reflection.ProcessorArchitecture;
+
+namespace Microsoft.Build.Tasks.AssemblyFoldersFromConfig
+{
+    internal class AssemblyFoldersFromConfig : IEnumerable<AssemblyFoldersFromConfigInfo>
+    {
+        /// <summary>
+        /// The list of directory names found from the config file.
+        /// </summary>
+        private readonly List<AssemblyFoldersFromConfigInfo> _directoryNames = new List<AssemblyFoldersFromConfigInfo>();
+
+        /// <summary>
+        /// Construct.
+        /// </summary>
+        /// <param name="targetRuntimeVersion">The runtime version property from the project file.</param>
+        internal AssemblyFoldersFromConfig(string configFile, string targetRuntimeVersion, ProcessorArchitecture targetArchitecture)
+        {
+            if (!FileUtilities.FileExistsNoThrow(configFile))
+            {
+                return;
+            }
+
+            var collection = AssemblyFolderCollection.Load(configFile);
+            var assemblyTargets = GatherVersionStrings(targetRuntimeVersion, collection);
+
+            bool is64BitOs = Environment.Is64BitOperatingSystem;
+            bool targeting64Bit = targetArchitecture == ProcessorArchitecture.Amd64 ||
+                                  targetArchitecture == ProcessorArchitecture.IA64;
+
+            // Platform-agnostic folders first.
+            FindDirectories(assemblyTargets, target => string.IsNullOrEmpty(target.Platform));
+
+            if (is64BitOs)
+            {
+                if (targeting64Bit)
+                {
+                    FindDirectories(assemblyTargets,
+                        target => !string.IsNullOrEmpty(target.Platform) && target.Platform.Equals("x64", StringComparison.OrdinalIgnoreCase));
+                    FindDirectories(assemblyTargets,
+                        target => !string.IsNullOrEmpty(target.Platform) && target.Platform.Equals("x86", StringComparison.OrdinalIgnoreCase));
+                }
+                else
+                {
+                    FindDirectories(assemblyTargets,
+                        target => !string.IsNullOrEmpty(target.Platform) && target.Platform.Equals("x86", StringComparison.OrdinalIgnoreCase));
+                    FindDirectories(assemblyTargets,
+                        target => !string.IsNullOrEmpty(target.Platform) && target.Platform.Equals("x64", StringComparison.OrdinalIgnoreCase));
+                }
+            }
+            else
+            {
+                FindDirectories(assemblyTargets,
+                    target => !string.IsNullOrEmpty(target.Platform) && target.Platform.Equals("x86", StringComparison.OrdinalIgnoreCase));
+            }
```

Comment:
I'll add a comment. But no I don't think it can be simplified that much because there are really 3 cases:
64bit and targeting 64 bit
64bit and targeting 32 bit
32bit

Those scenarios add folders in a different order.